File: iosApp/iosApp/Modules/NFTsDashboard/NFTsDashboardInteractor.swift
// Created by web3d4v on 24/05/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3walletcore

protocol NFTsDashboardInteractorLister: AnyObject {
    func popToRootAndRefresh()
}

protocol NFTsDashboardInteractor: AnyObject {
    func fetchYourNFTs(
        isPullDownToRefreh: Bool,
        onCompletion: @escaping (Result<[NFTItem], Error>) -> Void
    )
    func yourCollections() -> [NFTCollection]
    func addListener(_ listener: NFTsDashboardInteractorLister)
    func removeListener(_ listener: NFTsDashboardInteractorLister)
}

final class DefaultNFTsDashboardInteractor {
    private let networksService: NetworksService
    private let service: NFTsService
    
    private var listener: WeakContainer?

    init(
        networksService: NetworksService,
        service: NFTsService
    ) {
        self.networksService = networksService
        self.service = service
    }
}

extension DefaultNFTsDashboardInteractor: NFTsDashboardInteractor {
    
    func fetchYourNFTs(
        isPullDownToRefreh: Bool,
        onCompletion: @escaping (Result<[NFTItem], Error>) -> Void
    ) {
        if !isPullDownToRefreh {
            onCompletion(.success(service.yourNFTs()))
            return
        }
        service.fetchNFTs { items, error in
            if let error = error { onCompletion(.failure(error)) }
            else { onCompletion(.success(items ?? [])) }
        }
    }
    
    func yourCollections() -> [NFTCollection] { service.yourCollections() }
}

extension DefaultNFTsDashboardInteractor: NetworksListener {

    func addListener(_ listener: NFTsDashboardInteractorLister) {
        self.listener = WeakContainer(listener)
        networksService.add(listener__: self)
    }

    func removeListener(_ listener: NFTsDashboardInteractorLister) {
        self.listener = nil
        networksService.remove(listener__: self)
    }

    func handle(event_ event: NetworksEvent) {
        guard event is NetworksEvent.NetworkDidChange else { return }
        listener?.value?.popToRootAndRefresh()
    }

    private class WeakContainer {
        weak var value: NFTsDashboardInteractorLister?
        init(_ value: NFTsDashboardInteractorLister) {
            self.value = value
        }
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/nftsDashboard/NFTsDashboardInteractor.kt
package com.sonsofcrypto.web3walletcore.modules.nftsDashboard

import com.sonsofcrypto.web3lib.services.networks.NetworksEvent
import com.sonsofcrypto.web3lib.services.networks.NetworksListener
import com.sonsofcrypto.web3lib.services.networks.NetworksService
import com.sonsofcrypto.web3lib.utils.uiDispatcher
import com.sonsofcrypto.web3lib.utils.withUICxt
import com.sonsofcrypto.web3walletcore.services.nfts.NFTCollection
import com.sonsofcrypto.web3walletcore.services.nfts.NFTItem
import com.sonsofcrypto.web3walletcore.services.nfts.NFTsService
import com.sonsofcrypto.web3walletcore.services.nfts.NFTsServiceListener
import com.sonsofcrypto.web3walletcore.services.settings.NFTCarouselSize
import com.sonsofcrypto.web3walletcore.services.settings.NFTCarouselSize.*
import com.sonsofcrypto.web3walletcore.services.settings.SettingsService
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch

interface NFTsDashboardInteractorLister {
    fun networkChanged()
    fun nftsChanged()
}

interface NFTsDashboardInteractor {
    @Throws(Throwable::class)
    suspend fun fetchYourNFTs()
    fun nfts(): List<NFTItem>
    fun collections(): List<NFTCollection>
    fun regularCarousel(): Boolean
    fun add(listener: NFTsDashboardInteractorLister)
    fun remove(listener: NFTsDashboardInteractorLister)
}

class DefaultNFTsDashboardInteractor(
    private val networksService: NetworksService,
    private val nftsService: NFTsService,
    private val settingsService: SettingsService,
): NFTsDashboardInteractor, NetworksListener, NFTsServiceListener {
    private var nfts: List<NFTItem> = emptyList()
    private var collections: List<NFTCollection> = emptyList()
    private var listener: NFTsDashboardInteractorLister? = null

    override suspend fun fetchYourNFTs() {
        nftsService.fetchNFTs()
        withUICxt {
            nfts = nftsService.nfts()
            collections = nftsService.collections()
        }
    }

    override fun nfts(): List<NFTItem> = nfts
    override fun collections(): List<NFTCollection> = collections
    override fun regularCarousel(): Boolean =
        settingsService.nftCarouselSize == REGULAR

    override fun add(listener: NFTsDashboardInteractorLister) {
        if (this.listener != null) { remove(this.listener!!)}
        this.listener = listener
        networksService.add(this)
        nftsService.addListener(this)
    }

    override fun remove(listener: NFTsDashboardInteractorLister) {
        this.listener = null
        networksService.remove(this)
        nftsService.removeListener(this)
    }

    override fun handle(event: NetworksEvent) {
        if (event is NetworksEvent.NetworkDidChange) listener?.networkChanged()
    }

    override fun nftsChanged() {
        CoroutineScope(uiDispatcher).launch {
            nfts = nftsService.nfts()
            collections = nftsService.collections()
            listener?.nftsChanged()
        }
    }
}
